cunumeric/deferred.py

```diff
@@ -305,37 +305,288 @@ def get_scalar_array(self):
+                    transpose_indices += (dim,)
+                    transpose_needed = transpose_needed or (
+                        (dim - last_index) > 1
+                    )
+                    last_index = dim
+
+            if transpose_needed:
+                copy_needed = True
+                start_index = 0
+                post_indices = tuple(
+                    i for i in range(store.ndim) if i not in transpose_indices
+                )
+                transpose_indices += post_indices
+                store = store.transpose(transpose_indices)
+                index_map = list(transpose_indices)
+                count = 0
+                for i in transpose_indices:
+                    index_map[i] = count
+                    count += 1
+            else:
+                index_map = tuple(range(len(key)))
+
+            for d, k in enumerate(key):
+                dim = index_map[d]
+                if np.isscalar(k):
+                    if k < 0:
+                        k += store.shape[dim + shift]
+                    store = store.project(dim + shift, k)
+                    shift -= 1
+                    copy_needed = True
+                elif k is np.newaxis:
+                    store = store.promote(dim + shift, 1)
+                    copy_needed = True
+                elif isinstance(k, slice):
+                    store = store.slice(dim + shift, k)
+                    if k != slice(None):
+                        copy_needed = True
+                elif isinstance(k, NumPyThunk):
+                    if k.dtype == np.bool:
+                        if k.shape[0] != store.shape[dim]:
+                            raise ValueError(
+                                "boolean index did not match "
+                                "indexed array along dimension  "
+                            )
+                        # in case of the mixed indises we all nonzero
+                        # for the bool array
+                        k = k.nonzero()
+                        tuple_of_arrays += k
+                    else:
+                        tuple_of_arrays += (self.runtime.to_deferred_array(k),)
+                else:
+                    raise TypeError(
+                        "Unsupported entry type passed to advanced",
+                        "indexing operation",
                     )
-                tuple_of_arrays += (k,)
+            if copy_needed or (not store._transform.bottom):
+                # after store is transformed we need to to return a copy of
+                # the store since Copy operation can't be done on
+                # the store with transformation
+                rhs, store = self.copy_store(store)
         else:
             assert isinstance(key, NumPyThunk)
+            if not store._transform.bottom:
+                rhs, store = self.copy_store(store)
+            # the use case when index array ndim >1 and input array ndim ==1
```

Comment:
Where does this show up? It seems to be disallowed in NumPy:
```
>>> x = np.arange(4)
>>> x[ np.array([[True,False,True,False],[False,True,False,True]]) ]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed
```

Reply:
I see, this is talking about the integer case, not the boolean case. However, this check is too strict, e.g. it rejects this example, which is valid in NumPy:

```
x = cn.ones((3,4))
ind = cn.ones((2,2,2), dtype=int)
print(x[ind].shape)
```

Reply:
yes, I will remove this check